Code/Tools/AssetProcessor/Platform/Linux/native/FileWatcher/FileWatcher_linux.cpp

```diff
@@ -20,150 +23,151 @@ static constexpr size_t s_iNotifyMaxEntries = 1024 * 16;         // Control the
+            inotify_rm_watch(m_iNotifyHandle, watchHandle);
         }
+        m_handleToFolderMap.clear();
+        m_handleToFolderMapLock.unlock();
+
+        ::close(m_iNotifyHandle);
+        m_iNotifyHandle = -1;
     }
+}
 
-    void AddWatchFolder(QString folder)
+void FileWatcher::PlatformImplementation::AddWatchFolder(QString folder, bool recursive)
+{
+    if (m_iNotifyHandle < 0)
     {
-        if (m_iNotifyHandle >= 0)
-        {
-            // Clean up the path before accepting it as a watch folder
-            QString cleanPath = QDir::cleanPath(folder);
-
-            // Add the folder to watch and track it
-            int watchHandle = inotify_add_watch(m_iNotifyHandle, 
-                                                cleanPath.toUtf8().constData(),
-                                                IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MODIFY | IN_MOVE);
-            
-            if (!m_handleToFolderMapLock.tryLock(s_handleToFolderMapLockTimeout))
-            {
-                AZ_Error("FileWatcher", false, "Unable to obtain inotify handle lock on thread");
-                return;
-            }
-            m_handleToFolderMap[watchHandle] = cleanPath;
-            m_handleToFolderMapLock.unlock();
+        return;
+    }
 
-            // Add all the subfolders to watch and track them
-            QDirIterator dirIter(folder, QDirIterator::Subdirectories | QDirIterator::FollowSymlinks);
+    // Clean up the path before accepting it as a watch folder
+    QString cleanPath = QDir::cleanPath(folder);
 
-            while (dirIter.hasNext())
-            {
-                QString dirName = dirIter.next();
-                if (dirName.endsWith("/.") || dirName.endsWith("/.."))
-                {
-                    continue;
-                }
-                
-                int watchHandle = inotify_add_watch(m_iNotifyHandle, 
-                                                    dirName.toUtf8().constData(),
-                                                    IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MODIFY | IN_MOVE);
+    // Add the folder to watch and track it
+    int watchHandle = inotify_add_watch(m_iNotifyHandle,
+                                        cleanPath.toUtf8().constData(),
+                                        IN_CREATE | IN_CLOSE_WRITE | IN_DELETE | IN_DELETE_SELF | IN_MODIFY | IN_MOVE);
 
-                if (!m_handleToFolderMapLock.tryLock(s_handleToFolderMapLockTimeout))
-                {
-                    AZ_Error("FileWatcher", false, "Unable to obtain inotify handle lock on thread");
-                    return;
-                }
-                m_handleToFolderMap[watchHandle] = dirName;
-                m_handleToFolderMapLock.unlock();
-            }
-        }
+    if (watchHandle < 0)
+    {
+        AZStd::fixed_string<255> errorBuffer;
+        const char* errorString = strerror_r(errno, errorBuffer.data(), errorBuffer.size());
```

Comment:
The errorBuffer size here will be incorrect I believe. We may want to resize it as you did earlier in the PR so that if we add ASAN support to the containers this won't be flagged as a write to invalid memory.

Reply:
@jeremyong-az Is correct about this?
You should use `errorBuffer.capacity()` and then use `errorBuffer.resize_no_construct(strlen(errorString));` to fix the errorBuffer size and set the null-terminating character